server/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java

```diff
@@ -112,24 +114,25 @@ public WaitForInitialState(TransportMasterNodeAction<Request, Response>.AsyncSin
         @Override
         public void clusterChanged(ClusterChangedEvent event) {
             clusterService.removeListener(this);
-            asyncSingleAction.start();
+            asyncSingleAction.doStart(event.state());
         }
     }
 
     class AsyncSingleAction {
 
         private final ActionListener<Response> listener;
         private final Request request;
-        private volatile ClusterStateObserver observer;
+        private ClusterStateObserver observer;
+        private final long startTime;
 
         AsyncSingleAction(Request request, ActionListener<Response> listener) {
             this.request = request;
             this.listener = listener;
+            this.startTime = threadPool.relativeTimeInMillis();
         }
 
-        public void start() {
-            ClusterState state = clusterService.state();
-            if (state == null) {
+        protected void doStart(ClusterState clusterState) {
+            if (clusterState == null) {
```

Comment:
since we moved former `start()` into `doStart()` I moved there a [fix](https://github.com/crate/crate/pull/11702/) as well.